source/slang/slang-check-expr.cpp

```diff
@@ -2068,7 +2068,19 @@ IntVal* SemanticsVisitor::tryConstantFoldExpr(
             return nullptr;
         if (!isValidCompileTimeConstantType(substType))
             return nullptr;
-        auto val = tryConstantFoldExpr(typeCastOperand, kind, circularityInfo);
+
+        IntVal* val = nullptr; 
+        if (auto floatLitExpr = typeCastOperand.as<FloatingPointLiteralExpr>())
+        {
+            // When explicitly casting from float type to integer type, let's fold it as an integer value.
+            const IntegerLiteralValue value = IntegerLiteralValue(floatLitExpr.getExpr()->value);
+            val = m_astBuilder->getIntVal(m_astBuilder->getFloatType(), value);
+        }
+        else
+        {
+            val = tryConstantFoldExpr(typeCastOperand, kind, circularityInfo);
```

Comment:
Will this lead to infinite recursion?

Reply:
I don't see how it can become infinite recursion.

The function, tryConstantFoldExpr, is recursively called. But it peels one step for each call.

And that is not the code change I am making.
My change only adds a line to handle the case where the value is a floating-point literal.

Reply:
Ah I see, yeah this should work.